File: NowDepartingAndroid/wear/src/main/java/com/move38/nowdeparting/data/api/MTAFeedService.kt
package com.move38.nowdeparting.data.api

import android.location.Location
import android.util.Log
import com.move38.nowdeparting.data.model.Station
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.async
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.withContext
import okhttp3.OkHttpClient
import okhttp3.Request
import java.time.Instant
import java.util.concurrent.ConcurrentHashMap
import javax.inject.Inject
import javax.inject.Singleton

data class MTANearbyArrival(
    val routeId: String,
    val gtfsStopId: String?,
    val stationName: String,
    val stationDisplay: String,
    val direction: String,
    val arrivalTime: Instant,
    val distanceInMeters: Double
)

sealed class MTAFeedError : Exception() {
    data class MissingStopId(val stationName: String) : MTAFeedError()
    data class NetworkError(override val cause: Throwable) : MTAFeedError()
    data class HttpError(val code: Int) : MTAFeedError()
    data class ParseError(override val cause: Throwable) : MTAFeedError()
    object NoData : MTAFeedError()
}

@Singleton
class MTAFeedService @Inject constructor(
    private val okHttpClient: OkHttpClient
) {
    companion object {
        private const val TAG = "MTAFeedService"
        private const val CACHE_TTL_MS = 30_000L  // GTFS-RT feeds refresh ~every 30s
    }

    private val parser = GTFSRTParser()
    // Cache: feed URL → (fetchTimeMs, parsedUpdates)
    private val cache = ConcurrentHashMap<String, Pair<Long, List<GTFSTripUpdate>>>()

    // MARK: - By-Station Query

    /** Fetches arrival times at a specific station for a given route and direction. */
    suspend fun fetchArrivals(
        routeId: String,
        station: Station,
        direction: String
    ): Result<List<Instant>> = withContext(Dispatchers.IO) {
        val parentStopId = station.gtfsStopId
        if (parentStopId.isNullOrEmpty()) {
            return@withContext Result.failure(MTAFeedError.MissingStopId(station.name))
        }

        val targetStopId = parentStopId + direction  // e.g. "127" + "N" → "127N"

        fetchFeed(routeId).map { updates ->
            val now = Instant.now()
            updates
                .filter { it.routeId == routeId }
                .flatMap { update ->
                    update.stopTimeUpdates
                        .filter { it.stopId == targetStopId }
                        .mapNotNull { it.arrivalTime ?: it.departureTime }
                }
                .filter { it.isAfter(now) }
                .sorted()
        }
    }

    // MARK: - By-Location Query

    /** Fetches all arrivals within radiusMeters of the given location across every
     *  feed group. Returns raw MTANearbyArrival values. */
    suspend fun fetchNearbyArrivals(
        latitude: Double,
        longitude: Double,
        radiusMeters: Float = 1600f,
        stationsByLine: Map<String, List<Station>>
    ): Result<List<MTANearbyArrival>> = withContext(Dispatchers.IO) {
        data class NearbyStop(val station: Station, val distance: Float)

        val nearbyStops = mutableListOf<NearbyStop>()
        val seenNames = mutableSetOf<String>()

        for (stations in stationsByLine.values) {
            for (station in stations) {
                if (!seenNames.add(station.name)) continue
                station.gtfsStopId ?: continue
                val lat = station.latitude ?: continue
                val lon = station.longitude ?: continue

                val distResults = FloatArray(1)
                Location.distanceBetween(latitude, longitude, lat, lon, distResults)
                if (distResults[0] <= radiusMeters) {
                    nearbyStops.add(NearbyStop(station, distResults[0]))
                }
            }
        }

        if (nearbyStops.isEmpty()) return@withContext Result.success(emptyList())

        val feedUrls = MTAFeedConfiguration.feedUrls(stationsByLine.keys.toList())

        coroutineScope {
            val feedResults = feedUrls.map { url ->
                async { fetchFeedData(url) }
            }.map { it.await() }

            val allUpdates = mutableListOf<GTFSTripUpdate>()
            var firstError: MTAFeedError? = null

            for (result in feedResults) {
                result.fold(
                    onSuccess = { allUpdates.addAll(it) },
                    onFailure = { if (firstError == null) firstError = it as? MTAFeedError }
                )
            }

            if (allUpdates.isEmpty() && firstError != null) {
                return@coroutineScope Result.failure(firstError!!)
            }

            val now = Instant.now()
            val results = mutableListOf<MTANearbyArrival>()

            for (stop in nearbyStops) {
                for (dir in listOf("N", "S")) {
                    val parentId = stop.station.gtfsStopId ?: continue
                    val stopId = parentId + dir

                    for (update in allUpdates) {
                        for (stu in update.stopTimeUpdates) {
                            if (stu.stopId != stopId) continue
                            val arrivalTime = stu.arrivalTime ?: stu.departureTime ?: continue
                            if (!arrivalTime.isAfter(now)) continue
                            val minutesAway = (arrivalTime.epochSecond - now.epochSecond) / 60
                            if (minutesAway < 0 || minutesAway > 30) continue

                            results.add(
                                MTANearbyArrival(
                                    routeId = update.routeId,
                                    gtfsStopId = stop.station.gtfsStopId,
                                    stationName = stop.station.name,
                                    stationDisplay = stop.station.display,
                                    direction = dir,
                                    arrivalTime = arrivalTime,
                                    distanceInMeters = stop.distance.toDouble()
                                )
                            )
                        }
                    }
                }
            }

            // Sort: primarily by arrival time; secondarily by distance when times
            // are within 1 minute of each other.
            val sorted = results.sortedWith { a, b ->
                val tA = a.arrivalTime.epochSecond - now.epochSecond
                val tB = b.arrivalTime.epochSecond - now.epochSecond
                if (Math.abs(tA - tB) < 60) a.distanceInMeters.compareTo(b.distanceInMeters)
                else tA.compareTo(tB)
            }

            Result.success(sorted)
        }
    }

    // MARK: - Availability Check

    /** Checks which stations on a line have live arrivals and updates hasAvailableTimes. */
    suspend fun checkAvailability(
        lineId: String,
        stations: List<Station>
    ): List<Station> = withContext(Dispatchers.IO) {
        val result = fetchFeed(lineId)
        if (result.isFailure) return@withContext stations

        val updates = result.getOrDefault(emptyList())
        val activeStopIds = updates
            .filter { it.routeId == lineId }
            .flatMapTo(mutableSetOf()) { update ->
                update.stopTimeUpdates.map { it.stopId }
            }

        stations.map { station ->
            val parentId = station.gtfsStopId
            val hasData = if (parentId != null) {
                activeStopIds.contains(parentId + "N") || activeStopIds.contains(parentId + "S")
            } else {
                false
            }
            station.copy(hasAvailableTimes = hasData)
        }
    }

    // MARK: - Feed Fetch + Cache

    private suspend fun fetchFeed(routeId: String): Result<List<GTFSTripUpdate>> {
        val url = MTAFeedConfiguration.feedUrl(routeId)
            ?: return Result.failure(
                MTAFeedError.NetworkError(IllegalArgumentException("Unknown route: $routeId"))
            )
        return fetchFeedData(url)
    }

    /** Fetches and parses a GTFS-RT feed, using the in-memory cache when fresh.
     *  Must be called from a background thread (uses blocking OkHttp execute()). */
    private fun fetchFeedData(url: String): Result<List<GTFSTripUpdate>> {
        val cached = cache[url]
        if (cached != null && System.currentTimeMillis() - cached.first < CACHE_TTL_MS) {
            return Result.success(cached.second)
        }

        return try {
            val request = Request.Builder().url(url).build()
            val response = okHttpClient.newCall(request).execute()

            if (!response.isSuccessful) {
                return Result.failure(MTAFeedError.HttpError(response.code))
            }

            val bytes = response.body?.bytes()
            if (bytes == null || bytes.isEmpty()) {
                return Result.failure(MTAFeedError.NoData)
            }

            val updates = parser.parse(bytes)
            cache[url] = Pair(System.currentTimeMillis(), updates)
            Result.success(updates)
        } catch (e: GTFSRTParser.ParseError) {
            Result.failure(MTAFeedError.ParseError(e))
        } catch (e: Exception) {
            Log.e(TAG, "Error fetching feed: $url", e)
            Result.failure(MTAFeedError.NetworkError(e))
        }
    }
}

File: Shared/MTAFeedService.swift
//
//  MTAFeedService.swift
//  Now Departing
//
//  High-level service that replaces the third-party api.wheresthefuckingtrain.com
//  JSON API with direct calls to the MTA's official GTFS-RT binary feeds.
//
//  ## How it works
//
//  1.  The MTA publishes GTFS-RT "trip update" feeds (one per group of lines).
//      Each feed is a binary Protocol Buffer file containing real-time arrival
//      predictions for every in-service train in that group.
//
//  2.  `MTAFeedService` fetches the relevant feed(s), parses them with
//      `GTFSRTParser`, and returns the subset of arrivals the caller wants.
//
//  3.  GTFS-RT identifies stations by *stop ID*, e.g. "127N" for the
//      northbound 1/2/3 platform at Times Square.  To translate between the
//      user-visible station names stored in `stations.json` and GTFS stop IDs,
//      each `Station` now carries an optional `gtfsStopId` field (the *parent*
//      station ID, without the N/S direction suffix).
//
//      To populate `gtfsStopId` values in `stations.json`, run the helper
//      script at `scripts/generate_gtfs_mapping.py` once against the MTA's
//      GTFS static data.  See that script for full instructions.
//
//  ## Station-lookup approach
//
//  •  For **by-station** queries (`TimesViewModel`):
//     Use `Station.gtfsStopId` + direction suffix ("N" or "S") to find the
//     matching stop in the parsed feed.
//
//  •  For **nearby-train** queries (`NearbyTrainsManager`):
//     Use `Station.location` (lat/lon from `stations.json`) to compute the
//     distance from the user, then pull arrivals for all stops within the
//     desired radius from every feed group.
//

import Foundation
import CoreLocation

// MARK: - Nearby Arrival (target-agnostic intermediate type)

/// Raw result from `fetchNearbyArrivals`.  Converted to `NearbyTrain` by
/// `NearbyTrainsManager`, which is the only target that needs the richer type.
struct MTANearbyArrival {
    let routeId: String
    let gtfsStopId: String?
    let stationName: String
    let stationDisplay: String
    let direction: String    // "N" or "S"
    let arrivalTime: Date
    let distanceInMeters: Double
    let latitude: Double?
    let longitude: Double?
}


enum MTAFeedError: Error, LocalizedError {
    case missingStopId(stationName: String)
    case networkError(Error)
    case httpError(Int)
    case parseError(Error)
    case noData

    var errorDescription: String? {
        switch self {
        case .missingStopId(let name):
            return "No GTFS stop ID found for station \"\(name)\". Update stations.json."
        case .networkError(let e):
            return "Network error: \(e.localizedDescription)"
        case .httpError(let code):
            return "MTA server returned HTTP \(code)."
        case .parseError(let e):
            return "Failed to parse GTFS-RT feed: \(e.localizedDescription)"
        case .noData:
            return "No data received from MTA feed."
        }
    }
}

// MARK: - Service

/// Fetches and parses MTA GTFS-RT feeds, returning arrival times compatible
/// with the existing `TimesViewModel` and `NearbyTrainsManager` interfaces.
final class MTAFeedService {

    static let shared = MTAFeedService()

    private let parser = GTFSRTParser()
    private let session: URLSession

    // Simple in-memory cache: feed URL → (fetchDate, parsedUpdates)
    private var cache: [URL: (date: Date, updates: [GTFSTripUpdate])] = [:]
    private var alertCache: [URL: (date: Date, alerts: [GTFSAlert])] = [:]
    private let cacheTTL: TimeInterval = 30  // seconds; GTFS-RT feeds refresh ~every 30s

    private init() {
        let config = URLSessionConfiguration.default
        config.timeoutIntervalForRequest = 30
        config.timeoutIntervalForResource = 30
        session = URLSession(configuration: config)
    }

    // MARK: - By-Station Query (replaces /by-route/<lineId>)

    /// Fetches arrival times at a specific station for a given route and direction.
    ///
    /// - Parameters:
    ///   - routeId:     Subway route ID, e.g. "1", "A", "L".
    ///   - station:     The `Station` value from `stations.json`.  Its
    ///                  `gtfsStopId` must be non-nil for the lookup to work.
    ///   - direction:   "N" (northbound / uptown) or "S" (southbound / downtown).
    ///   - completion:  Called on the main thread with arrival `Date` values or an error.
    func fetchArrivals(
        routeId: String,
        station: Station,
        direction: String,
        completion: @escaping (Result<[Date], MTAFeedError>) -> Void
    ) {
        guard let parentStopId = station.gtfsStopId, !parentStopId.isEmpty else {
            DispatchQueue.main.async {
                completion(.failure(.missingStopId(stationName: station.name)))
            }
            return
        }

        let targetStopId = parentStopId + direction  // e.g. "127" + "N" → "127N"

        fetchFeed(for: routeId) { result in
            switch result {
            case .failure(let error):
                DispatchQueue.main.async { completion(.failure(error)) }

            case .success(let updates):
                let now = Date()
                let arrivals: [Date] = updates
                    .filter { $0.routeId == routeId }
                    .flatMap { update -> [Date] in
                        update.stopTimeUpdates
                            .filter { $0.stopId == targetStopId }
                            .compactMap { $0.arrivalTime ?? $0.departureTime }
                    }
                    .filter { $0 > now }
                    .sorted()

                DispatchQueue.main.async { completion(.success(arrivals)) }
            }
        }
    }

    // MARK: - By-Location Query (replaces /by-location?lat=…&lon=…)

    /// Fetches all arrivals within `radiusMeters` of `location` across every
    /// feed group and returns raw `MTANearbyArrival` values.
    /// `NearbyTrainsManager` converts these to the richer `NearbyTrain` type.
    ///
    /// - Parameters:
    ///   - location:       The user's current `CLLocation`.
    ///   - radiusMeters:   Search radius (default 800 m ≈ 0.5 miles).
    ///   - stationsByLine: The full station dictionary from `StationDataManager`.
    ///   - completion:     Called on the main thread.
    func fetchNearbyArrivals(
        location: CLLocation,
    radiusMeters: Double = 1600,    // ~ 1 mile radius
        stationsByLine: [String: [Station]],
        completion: @escaping (Result<[MTANearbyArrival], MTAFeedError>) -> Void
    ) {
        // Collect all unique stations that have a GTFS stop ID and a known location,
        // and are within the search radius.
        struct NearbyStop {
            let station: Station
            let distance: Double
            let stationLocation: CLLocation
        }

        var nearbyStops: [NearbyStop] = []
        var seenStationNames = Set<String>()

        for stations in stationsByLine.values {
            for station in stations {
                guard !seenStationNames.contains(station.name),
                      station.gtfsStopId != nil,
                      let lat = station.latitude,
                      let lon = station.longitude else { continue }

                let stationLocation = CLLocation(latitude: lat, longitude: lon)
                let distance = location.distance(from: stationLocation)
                if distance <= radiusMeters {
                    nearbyStops.append(NearbyStop(station: station, distance: distance, stationLocation: stationLocation))
                    seenStationNames.insert(station.name)
                }
            }
        }

        guard !nearbyStops.isEmpty else {
            DispatchQueue.main.async { completion(.success([MTANearbyArrival]())) }
            return
        }

        // Determine which GTFS-RT feeds we need.
        let allRoutes = stationsByLine.keys.map { $0 }
        let feedURLs = MTAFeedConfiguration.feedURLs(for: allRoutes)

        // Fetch all required feeds concurrently.
        let group = DispatchGroup()
        var allUpdates: [GTFSTripUpdate] = []
        var firstError: MTAFeedError?
        let lock = NSLock()

        for url in feedURLs {
            group.enter()
            fetchFeedData(url: url) { result in
                defer { group.leave() }
                switch result {
                case .success(let updates):
                    lock.lock()
                    allUpdates.append(contentsOf: updates)
                    lock.unlock()
                case .failure(let error):
                    lock.lock()
                    if firstError == nil { firstError = error }
                    lock.unlock()
                }
            }
        }

        group.notify(queue: .global(qos: .userInitiated)) {
            if allUpdates.isEmpty, let error = firstError {
                DispatchQueue.main.async { completion(.failure(error)) }
                return
            }

            let now = Date()
            var results: [MTANearbyArrival] = []

            for stop in nearbyStops {
                for direction in ["N", "S"] {
                    guard let parentId = stop.station.gtfsStopId else { continue }
                    let stopId = parentId + direction

                    let arrivals: [(route: String, time: Date)] = allUpdates
                        .flatMap { update -> [(String, Date)] in
                            update.stopTimeUpdates
                                .filter { $0.stopId == stopId }
                                .compactMap { stu -> (String, Date)? in
                                    guard let t = stu.arrivalTime ?? stu.departureTime else { return nil }
                                    return (update.routeId, t)
                                }
                        }
                        .filter { $0.1 > now }

                    for (routeId, arrivalTime) in arrivals {
                        let minutesAway = Int(arrivalTime.timeIntervalSinceNow / 60)
                        guard minutesAway >= 0, minutesAway <= 30 else { continue }

                        results.append(MTANearbyArrival(
                            routeId: routeId,
                            gtfsStopId: stop.station.gtfsStopId,
                            stationName: stop.station.name,
                            stationDisplay: stop.station.display,
                            direction: direction,
                            arrivalTime: arrivalTime,
                            distanceInMeters: stop.distance,
                            latitude: stop.station.latitude,
                            longitude: stop.station.longitude
                        ))
                    }
                }
            }

            // Sort: primarily by arrival time; secondarily by distance when times
            // are within 1 minute of each other (matches previous WTFT behaviour).
            let sorted = results.sorted { a, b in
                let tA = a.arrivalTime.timeIntervalSinceNow
                let tB = b.arrivalTime.timeIntervalSinceNow
                if abs(tA - tB) < 60 { return a.distanceInMeters < b.distanceInMeters }
                return tA < tB
            }

            DispatchQueue.main.async { completion(.success(sorted)) }
        }
    }

    // MARK: - Feed Availability Check (replaces /by-route/<lineId> in StationDataManager)

    /// Checks which stations on a line have live arrivals and updates
    /// `hasAvailableTimes` on each `Station`.
    func checkAvailability(
        for lineId: String,
        stations: [Station],
        completion: @escaping ([Station]) -> Void
    ) {
        fetchFeed(for: lineId) { result in
            guard case .success(let updates) = result else {
                DispatchQueue.main.async { completion(stations) }
                return
            }

            // Collect all stop_ids that appear in at least one update.
            let activeStopIds = Set(
                updates.filter { $0.routeId == lineId }
                       .flatMap { $0.stopTimeUpdates.map { $0.stopId } }
            )

            let updated: [Station] = stations.map { station in
                var s = station
                if let parentId = station.gtfsStopId {
                    let hasN = activeStopIds.contains(parentId + "N")
                    let hasS = activeStopIds.contains(parentId + "S")
                    s.hasAvailableTimes = hasN || hasS
                } else {
                    s.hasAvailableTimes = false
                }
                return s
            }

            DispatchQueue.main.async { completion(updated) }
        }
    }

    // MARK: - Service Alerts

    /// Fetches and parses the MTA GTFS-RT Alerts feed, returning all active service alerts.
    /// Results are cached for `cacheTTL` seconds.
    func fetchServiceAlerts(
        completion: @escaping (Result<[GTFSAlert], MTAFeedError>) -> Void
    ) {
        guard let url = MTAFeedConfiguration.alertsFeedURL else {
            completion(.failure(.networkError(URLError(.badURL))))
            return
        }

        // Return cached alerts if still fresh.
        if let cached = alertCache[url], Date().timeIntervalSince(cached.date) < cacheTTL {
            DispatchQueue.main.async { completion(.success(cached.alerts)) }
            return
        }

        let request = MTAFeedConfiguration.request(for: url)
        print("[MTAFeed] Requesting alerts URL: \(url)")
        session.dataTask(with: request) { [weak self] data, response, error in
            guard let self = self else { return }

            if let error = error {
                print("[MTAFeed] Alerts network error: \(error)")
                DispatchQueue.main.async { completion(.failure(.networkError(error))) }
                return
            }

            if let http = response as? HTTPURLResponse {
                print("[MTAFeed] Alerts HTTP status: \(http.statusCode)")
                if !(200...299).contains(http.statusCode) {
                    DispatchQueue.main.async { completion(.failure(.httpError(http.statusCode))) }
                    return
                }
            }

            guard let data = data, !data.isEmpty else {
                print("[MTAFeed] Alerts: no data received")
                DispatchQueue.main.async { completion(.failure(.noData)) }
                return
            }

            print("[MTAFeed] Alerts: received \(data.count) bytes, parsing…")
            do {
                let alerts = try self.parser.parseAlerts(data)
                print("[MTAFeed] Alerts: parsed \(alerts.count) alerts")
                self.alertCache[url] = (date: Date(), alerts: alerts)
                DispatchQueue.main.async { completion(.success(alerts)) }
            } catch {
                print("[MTAFeed] Alerts parse error: \(error)")
                DispatchQueue.main.async { completion(.failure(.parseError(error))) }
            }
        }.resume()
    }

    // MARK: - Feed Fetch + Cache

    /// Fetches and parses the GTFS-RT feed for a given route, using the cache
    /// when the cached data is still within `cacheTTL` seconds old.
    func fetchFeed(
        for routeId: String,
        completion: @escaping (Result<[GTFSTripUpdate], MTAFeedError>) -> Void
    ) {
        guard let url = MTAFeedConfiguration.feedURL(for: routeId) else {
            completion(.failure(.networkError(URLError(.badURL))))
            return
        }
        fetchFeedData(url: url, completion: completion)
    }

    private func fetchFeedData(
        url: URL,
        completion: @escaping (Result<[GTFSTripUpdate], MTAFeedError>) -> Void
    ) {
        // Return cached result if fresh enough.
        if let cached = cache[url], Date().timeIntervalSince(cached.date) < cacheTTL {
            completion(.success(cached.updates))
            return
        }

        let request = MTAFeedConfiguration.request(for: url)

        session.dataTask(with: request) { [weak self] data, response, error in
            guard let self = self else { return }

            if let error = error {
                completion(.failure(.networkError(error)))
                return
            }

            if let http = response as? HTTPURLResponse, !(200...299).contains(http.statusCode) {
                completion(.failure(.httpError(http.statusCode)))
                return
            }

            guard let data = data, !data.isEmpty else {
                completion(.failure(.noData))
                return
            }

            do {
                let updates = try self.parser.parse(data)
                self.cache[url] = (date: Date(), updates: updates)
                completion(.success(updates))
            } catch {
                completion(.failure(.parseError(error)))
            }
        }.resume()
    }
}
